libp2p/src/main/java/io/libp2p/protocol/circuit/RelayTransport.java

```diff
@@ -0,0 +1,334 @@
+    // first connect to relay over hop
+    List<MultiaddrComponent> comps = multiaddr.getComponents();
+    int split = comps.indexOf(new MultiaddrComponent(Protocol.P2PCIRCUIT, null));
+    Multiaddr relay = new Multiaddr(comps.subList(0, split));
+    Multiaddr target = new Multiaddr(comps.subList(split, comps.size()));
+    CircuitHopProtocol.HopController ctr = hop.dial(us, relay).getController().join();
+    // request proxy to target
+    Stream stream = ctr.connect(target.getPeerId()).join();
+    // upgrade with sec and muxer
+    return upgradeStream(stream, true, upgrader, this, target.getPeerId(), connHandler);
+  }
+
+  public static CompletableFuture<Connection> upgradeStream(
+      Stream stream,
+      boolean isInitiator,
+      ConnectionUpgrader upgrader,
+      Transport transport,
+      PeerId remote,
+      ConnectionHandler connHandler) {
+    ConnectionOverStream conn = new ConnectionOverStream(isInitiator, transport, stream);
+    CompletableFuture<Connection> res = new CompletableFuture<>();
+    stream.pushHandler(
+        new ChannelInitializer<>() {
+          @Override
+          protected void initChannel(Channel channel) throws Exception {
+            channel.attr(AttributesKt.getREMOTE_PEER_ID()).set(remote);
+            channel.attr(AttributesKt.getCONNECTION()).set(conn);
+            upgrader
+                .establishSecureChannel(conn)
+                .thenCompose(
+                    sess -> {
+                      conn.setSecureSession(sess);
+                      if (sess.getEarlyMuxer() != null) {
+                        return ConnectionUpgrader.Companion.establishMuxer(
+                            sess.getEarlyMuxer(), conn);
+                      } else {
+                        return upgrader.establishMuxer(conn);
+                      }
+                    })
+                .thenAccept(
+                    sess -> {
+                      conn.setMuxerSession(sess);
+                      connHandler.handleConnection(conn);
+                      res.complete(conn);
+                    })
+                .exceptionally(
+                    t -> {
+                      res.completeExceptionally(t);
+                      return null;
+                    });
+            channel.pipeline().fireChannelActive();
+          }
+        });
+    return res;
+  }
+
+  @Override
+  public boolean handles(@NotNull Multiaddr multiaddr) {
+    return multiaddr.hasAny(Protocol.P2PCIRCUIT);
+  }
+
+  @Override
+  public void initialize() {
+    stop.setTransport(this);
+    // find relays and connect and reserve
+    new Thread(
+            () -> {
+              while (true) {
+                ensureEnoughCurrentRelays();
+                try {
+                  Thread.sleep(2 * 60_000);
+                } catch (InterruptedException e) {
+                }
+              }
+            },
+            "RelayManager")
+        .start();
+  }
```

Comment:
I would use `ScheduledExecutorService` here which is explicitly passed via `RelayTransport` constructor